Document/0x06f-Testing-Local-Authentication.md

```diff
@@ -9,26 +9,196 @@ As a result only requirement "4.6	Biometric authentication, if any, is not event
+query[kSecUseOperationPrompt as String] = "Please, pass authorisation to enter this area" as CFString
+
+// 2. get item
+var queryResult: AnyObject?
+let status = withUnsafeMutablePointer(to: &queryResult) {
+    SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+}
+
+if status == noErr {
+    let password = String(data: queryResult as! Data, encoding: .utf8)!
+    // successfully received password
+} else {
+    // authorization not passed
+}
+```
+###### Objective-C
+```
+// 1. define query
+NSDictionary *query = @{(__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecReturnData: @YES,
+    (__bridge id)kSecAttrAccount: @"My Name1",
+    (__bridge id)kSecAttrLabel: @"com.me.myapp.password",
+    (__bridge id)kSecUseOperationPrompt: @"Please, pass authorisation to enter this area" };
+
+// 2. get item
+CFTypeRef queryResult = NULL;
+OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &queryResult);
+
+if (status == noErr){
+    NSData *resultData = ( __bridge_transfer NSData *)queryResult;
+    NSString *password = [[NSString alloc] initWithData:resultData encoding:NSUTF8StringEncoding];
+    NSLog(@"%@", password);
+} else {
+    NSLog(@"Something went wrong");
+}
+```
+#### Without Source Code
+Usage of frameworks in app can be detected by analyzing app binary's list of shared dynamic libraries by running:
+ `$ otool -L <AppName>.app/<AppName>`
+
+If `LocalAuhentication.framework` was used in app, output will contain both of following lines (remember that `LocalAuhentication.framework` uses `Security.framework` under the hood), if `Security.framework` - only second one:
+```
+/System/Library/Frameworks/LocalAuthentication.framework/LocalAuthentication
+/System/Library/Frameworks/Security.framework/Security
+```
 
 #### Dynamic Analysis
 
--- TODO [Describe how to test for this issue "Testing Biometric Authentication" by running and interacting with the app. This can include everything from simply monitoring network traffic or aspects of the app’s behavior to code injection, debugging, instrumentation, etc.] --
-
-#### Remediation
+On a running app usage of TouchID authentication is quite obvious: at appropriate moment system-style alert asks user to put her finger on sensor or fallback to password (if allowed):
+![TouchID authentication UI](/Images/Chapters/0x6f/biometric_auth_try_again.jpg)
+![TouchID authentication UI with fallback to passcode](/Images/Chapters/0x6f/biometric_auth_try_again.jpg)
 
-Example: https://www.raywenderlich.com/92667/securing-ios-data-keychain-touch-id-1password
--- TODO [Describe the best practices that developers should follow to prevent this issue "Testing Biometric Authentication".] --
+On a jailbroken device contents of Keychain can be dumped and items' parameters inspected.
+-- TODO [Will items saved via `Security.framework` Access Control API have any specific parameter in Keychain db?]
 
```

Comment:
Maybe I missed it: but i guess it would be nice to change this a bit.
- static analysis: show what people should not do (NEVER do LAContext based obj.-c breaks as that does not help at all, try to circumvent LAContext based stuff in Swift, never use "userpresence" as a token in high risk applications, etc.)
- dynamic analysis: explain what you did, but rather in the smaller steps and make sure that you cannot bypass it
- remediation: always use a crypto or password protected mechanism with the right access-controls in both swift and objective-c